src/main/java/org/apache/commons/lang3/ArrayUtils.java

```diff
@@ -1868,6 +1868,1013 @@ public static void reverse(final short[] array, final int startIndexInclusive, f
+        if (array == null) {
+            return;
+        }
+        shift(array, 0, array.length, offset);
+    }
+
+    /**
+     * <p>Shifts the order of the given array.</p>
+     *
+     * <p>This method does nothing for a {@code null} input array.</p>
+     *
+     * @param array  the array to shift, may be {@code null}
+     */
+    public static void shift(final double[] array, int offset) {
+        if (array == null) {
+            return;
+        }
+        shift(array, 0, array.length, offset);
+    }
+
+    /**
+     * <p>Shifts the order of the given array.</p>
+     *
+     * <p>This method does nothing for a {@code null} input array.</p>
+     *
+     * @param array  the array to shift, may be {@code null}
+     */
+    public static void shift(final float[] array, int offset) {
+        if (array == null) {
+            return;
+        }
+        shift(array, 0, array.length, offset);
+    }
+
+    /**
+     * <p>Shifts the order of the given array.</p>
+     *
+     * <p>This method does nothing for a {@code null} input array.</p>
+     *
+     * @param array  the array to shift, may be {@code null}
+     */
+    public static void shift(final boolean[] array, int offset) {
+        if (array == null) {
+            return;
+        }
+        shift(array, 0, array.length, offset);
+    }
+
+    /**
+     * <p>
+     * Shifts the order of the given array in the given range.
+     * </p>
+     * 
+     * <p>
+     * This method does nothing for a {@code null} input array.
+     * </p>
+     * 
+     * @param array
+     *            the array to shift, may be {@code null}
+     * @param startIndexInclusive
+     *            the starting index. Undervalue (&lt;0) is promoted to 0, overvalue (&gt;array.length) results in no
+     *            change.
+     * @param endIndexExclusive
+     *            elements up to endIndex-1 are shiftd in the array. Undervalue (&lt; start index) results in no
+     *            change. Overvalue (&gt;array.length) is demoted to array length.
+     * @since 3.2
+     */
+    public static void shift(final boolean[] array, int startIndexInclusive, int endIndexExclusive, int offset) {
+        if (array == null) {
+            return;
+        }
+        if (startIndexInclusive >= array.length - 1 || endIndexExclusive <= 0) {
+            return;
+        }
+        if (startIndexInclusive < 0) {
+            startIndexInclusive = 0;
+        } 
+        if (endIndexExclusive >= array.length) {
+            endIndexExclusive = array.length;
+        }        
+        int n = endIndexExclusive - startIndexInclusive;
+        if (n <= 1) {
+            return;
+        }
+        offset %= n;
+        if (offset < 0) {
+            offset += n;
+        }
+        while (n > 1 && offset > 0) {
+            int n_offset = n - offset;
+            
+            if (offset > n_offset) {
+                swap(array, startIndexInclusive, startIndexInclusive + n - n_offset,  n_offset);
```

Comment:
Sorry for the delay in answering, but I was busy and I wanted to give a full explanation.

Let me calculate the complexity of the shift algorithm. Let Sh(n, k) be the problem of shifting k positions in an array of size n and Sw(k) be the problem of swapping k elements in an array. 

It is obvious that O(Sw(k)) = O(k).

It is also obvious that O(Sh(1, k)) = O(1), with k < 1. Also O(Sh(x, 0)) = O(1) - we'll use that later.

Now let's assume that O(Sh(n, k)) = O(n), whatever n, with k < n. I'll try to prove that O(Sh(n + 1, k')) = O(n), with k' < n + 1.

Analyzing the algorithm we have
O (Sh(n + 1, k')) = max(
- O(Sw(n + 1 - k')) + O(Sh(k', 2k' - n - 1)), if 2k' > n + 1
- O(Sw(k')) + O(Sh(n + 1 - k', k')), if 2k' < n + 1
- O(Sw(k')), if 2k' == n + 1

First case 2k' > n + 1
 O(Sh(n + 1, k')) = O(Sw(n + 1 - k')) + O(Sh(k', 2k' - n - 1)) = O(n) + O(Sh(k', 2k' - n - 1))
Because k' < n +1 => k' <= n and 2k' <= 2n => 2k' - n - 1 <= n - 1 => 2k' - n - 1 < n, then O (Sh(n + 1, k')) = O(n) + O(n) = O(n).

Second case 2k' < n + 1
O(Sh(n + 1, k')) = O(Sw(k')) + O(Sh(n + 1 - k', k')) = O(k') + O(Sh(n + 1 - k', k')) = O(n) + O(Sh(n + 1 - k', k')) 
Because 0 < k'   => n + 1 - k' < n + 1 => n + 1 - k' <= n and k' < (n + 1)/2 => k' <= n/2 => k' < n, then O(Sh(n + 1, k')) = O(n) + O(n) = O(n).

Third case 2k' == n + 1
O(Sh(n + 1, k')) = O(Sw(k')) = O(n)

So O(Sh(n + 1, k')) = O(n)  (qed).

Am I missing something?